### ISSUE TITLE
Volume remains attached even after we do 'Reset' of a vm

### ISSUE THREAD

**Issue by ashahi1** [vmware/docker-volume-vsphere#1638]
Steps to reproduce:
```
1. Created a volume
2. Ran a container with volume attached
3. While container is running, 'Reset' the vm
4. After vm comes back up, ```docker volume ls ``` still shows volume as attached.
5. Not able to remove the volume and we have a left-over volume.
```

Is the behavior in step - 4 expected?

If I 'Restart Guest OS' on a vm, volume gets detached after restart. So why is this behavior different in case of 'Reset' of a vm?

Logs have been attached at the end.

---------------------

Steps and their output are as follows:

1. Created a volume - volABC
```
root@photon-machine [ ~ ]# docker volume create -d vsphere:latest --name=volABC
volABC
root@photon-machine [ ~ ]#
root@photon-machine [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      TestVolSuspend@sharedVmfs-0
vsphere:latest      VolZZZ@sharedVmfs-0
vsphere:latest      volABC@sharedVmfs-0
root@photon-machine [ ~ ]#

[root@sc-rdops-vm01-dhcp-20-85:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume          Datastore     VMGroup   Capacity  Used  Filesystem  Policy  Disk Format  Attached-to   Access      Attach-as               Created By    Created Date
--------------  ------------  --------  --------  ----  ----------  ------  -----------  ------------  ----------  ----------------------  ------------  ------------------------
VolZZZ          sharedVmfs-0  _DEFAULT  100MB     0     N/A         N/A     thin         photon-VM2.3  read-write  independent_persistent  photon-VM2.3  Tue Jul 25 20:50:26 2017
TestVolSuspend  sharedVmfs-0  _DEFAULT  100MB     15MB  ext4        N/A     thin         photon-VM0.3  read-write  independent_persistent  photon-VM0.3  Wed Jul 26 20:10:58 2017
volABC          sharedVmfs-0  _DEFAULT  100MB     13MB  ext4        N/A     thin         detached      read-write  independent_persistent  photon-VM0.3  Wed Jul 26 21:26:05 2017

[root@sc-rdops-vm01-dhcp-20-85:~]

```
2. Ran a container with volume mounted.
```
root@photon-machine [ ~ ]# docker run -it --volume-driver=vsphere:latest -v volABC@sharedVmfs-0:/vol1 --name ubABC ubuntu
root@4f0d8f7fb8cd:/#
root@4f0d8f7fb8cd:/# ls
bin  boot  dev  etc  home  lib  lib64  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var  vol1
root@4f0d8f7fb8cd:/# df -h
Filesystem                                       Size  Used Avail Use% Mounted on
overlay                                          7.8G  750M  6.6G  10% /
tmpfs                                            2.0G     0  2.0G   0% /dev
tmpfs                                            2.0G     0  2.0G   0% /sys/fs/cgroup
/dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:1:0   93M  1.6M   85M   2% /vol1
/dev/root                                        7.8G  750M  6.6G  10% /etc/hosts
shm                                               64M     0   64M   0% /dev/shm
tmpfs                                            2.0G     0  2.0G   0% /sys/firmware
root@4f0d8f7fb8cd:/#
```
3. Reset the VM
4. After vm comes back up, ``docker volume inspect``` and admin cli volume ls still shows volume as attached.
```
root@photon-machine [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      TestVolSuspend@sharedVmfs-0
vsphere:latest      VolZZZ@sharedVmfs-0
vsphere:latest      volABC@sharedVmfs-0
root@photon-machine [ ~ ]#
root@photon-machine [ ~ ]#
root@photon-machine [ ~ ]# docker volume inspect TestVolSuspend@sharedVmfs-0
[
    {
        "Driver": "vsphere:latest",
        "Labels": null,
        "Mountpoint": "/mnt/vmdk/TestVolSuspend@sharedVmfs-0",
        "Name": "TestVolSuspend@sharedVmfs-0",
        "Options": {},
        "Scope": "global",
        "Status": {
            "access": "read-write",
            "attach-as": "independent_persistent",
            "attached to VM": "photon-VM0.3",
            "attachedVMDevice": {
                "ControllerPciSlotNumber": "160",
                "Unit": "0"
            },
            "capacity": {
                "allocated": "15MB",
                "size": "100MB"
            },
            "clone-from": "None",
            "created": "Wed Jul 26 20:10:58 2017",
            "created by VM": "photon-VM0.3",
            "datastore": "sharedVmfs-0",
            "diskformat": "thin",
            "fstype": "ext4",
            "status": "attached"
        }
    }
]
root@photon-machine [ ~ ]#

[root@sc-rdops-vm01-dhcp-20-85:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume          Datastore     VMGroup   Capacity  Used  Filesystem  Policy  Disk Format  Attached-to   Access      Attach-as               Created By    Created Date
--------------  ------------  --------  --------  ----  ----------  ------  -----------  ------------  ----------  ----------------------  ------------  ------------------------
VolZZZ          sharedVmfs-0  _DEFAULT  100MB     0     N/A         N/A     thin         photon-VM2.3  read-write  independent_persistent  photon-VM2.3  Tue Jul 25 20:50:26 2017
TestVolSuspend  sharedVmfs-0  _DEFAULT  100MB     15MB  ext4        N/A     thin         photon-VM0.3  read-write  independent_persistent  photon-VM0.3  Wed Jul 26 20:10:58 2017
volABC          sharedVmfs-0  _DEFAULT  100MB     15MB  ext4        N/A     thin         photon-VM0.3  read-write  independent_persistent  photon-VM0.3  Wed Jul 26 21:26:05 2017

[root@sc-rdops-vm01-dhcp-20-85:~]

```

Logs:
[docker-volume-vsphere.txt](https://github.com/vmware/docker-volume-vsphere/files/1178157/docker-volume-vsphere.txt)
[vmdk_ops.txt](https://github.com/vmware/docker-volume-vsphere/files/1178156/vmdk_ops.txt)

**Comment by govint** [comment]
Most likely reset isn't posting a powered-off notification and the VM listener doesn't remove the volume.

But when Docker restarts in the VM it will restart the container, didn't that happen?

**Comment by govint** [comment]
This is not a bug, post the reset, the service detached the volume but then docker seems to have restarted the container and hence the behavior. Logs on both service and plugin confirm this.

From the logs,
1. The VMListener did get the powered-off notice and detached the disk

07/26/17 20:44:22 80986 [VMChangeListener] [INFO   ] VM poweroff change found for photon-VM0.3
07/26/17 20:44:22 80986 [VMChangeListener] [INFO   ] backing disk name is  photon-VM0.3/photon-VM0.3.vmdk
07/26/17 20:44:22 80986 [VMChangeListener] [INFO   ] backing disk name is  dockvols/11111111-1111-1111-1111-111111111111/TestVolSuspend.vmdk
07/26/17 20:44:22 80986 [VMChangeListener] [INFO   ] Setting detach status for /vmfs/volumes/sharedVmfs-0/dockvols/11111111-1111-1111-1111-111111111111/TestVolSuspend.vmdk
07/26/17 20:44:22 80986 [VMChangeListener] [INFO   ] Disk detached /vmfs/volumes/sharedVmfs-0/dockvols/11111111-1111-1111-1111-111111111111/TestVolSuspend.vmdk

2. Volume was reported as detached
07/26/17 20:46:48 80986 [MainThread] [INFO   ] Started new thread : 599781738240 with target <function execRequestThread at 0x8ba425b9d8> and args (13, 85930, b'{"cmd":"get","details":{"Name":"Test
VolSuspend"},"version":"2"}')
07/26/17 20:46:48 80986 [Thread-650] [INFO   ] Auth DB /etc/vmware/vmdkops/auth-db is missing, allowing all access
07/26/17 20:46:48 80986 [Thread-650] [WARNING] Internal: Tenant name symlink not found for path /vmfs/volumes/sharedVmfs-0/dockvols/_DEFAULT
07/26/17 20:46:48 80986 [photon-VM0.3-sharedVmfs-0._DEFAULT.TestVolSuspend] [INFO   ] executeRequest 'get' completed with ret={'access': 'read-write', 'status': 'detached', 'attach-as': 'independen
t_persistent', 'created': 'Wed Jul 26 20:10:58 2017', 'capacity': {'allocated': '15MB', 'size': '100MB'}, 'clone-from': 'None', 'created by VM': 'photon-VM0.3', 'diskformat': 'thin', 'datastore': '
sharedVmfs-0', 'fstype': 'ext4'}

3.  Attach was received for the volume,

07/26/17 20:49:06 80986 [photon-VM0.3-sharedVmfs-0._DEFAULT.TestVolSuspend] [INFO   ] *** disk_attach: VMDK /vmfs/volumes/sharedVmfs-0/dockvols/11111111-1111-1111-1111-111111111111/TestVolSuspend.vmdk to VM 'photon-VM0.3' , bios uuid = 564d0ff0-9e81-8795-f220-0bc248c8c3eb, VC uuid=527e4840-14d2-0a59-0d0a-65e9d8997c2d)

4.  And reported as attached thereafter,

07/26/17 21:02:40 80986 [MainThread] [INFO   ] Started new thread : 599781738240 with target <function execRequestThread at 0x8ba425b9d8> and args (11, 86275, b'{"cmd":"get","details":{"Name":"TestVolSuspend"},"version":"2"}')
07/26/17 21:02:40 80986 [Thread-665] [INFO   ] Auth DB /etc/vmware/vmdkops/auth-db is missing, allowing all access
07/26/17 21:02:40 80986 [Thread-665] [WARNING] Internal: Tenant name symlink not found for path /vmfs/volumes/sharedVmfs-0/dockvols/_DEFAULT
07/26/17 21:02:40 80986 [photon-VM0.3-sharedVmfs-0._DEFAULT.TestVolSuspend] [INFO   ] executeRequest 'get' completed with ret={'status': 'attached', 'attached to VM': 'photon-VM0.3', 'created by VM': 'photon-VM0.3', 'diskformat': 'thin', 'datastore': 'sharedVmfs-0', 'fstype': 'ext4', 'access': 'read-write', 'created': 'Wed Jul 26 20:10:58 2017', 'capacity': {'allocated': '15MB', 'size': '100MB'}, 'clone-from': 'None', 'attachedVMDevice': {'Unit': '0', 'ControllerPciSlotNumber': '160'}, 'attach-as': 'independent_persistent'}


Plugin client logs also confirm that a mount was done,

2017-07-26 20:46:50.34525868 +0000 UTC [INFO] Successfully retrieved mountsmap=map[]
2017-07-26 20:46:50.345299664 +0000 UTC [INFO] Discovered 0 volumes in use.
2017-07-26 20:46:50.345309652 +0000 UTC [INFO] Refcounting successfully completed
2017-07-26 20:49:07.31829256 +0000 UTC [INFO] Mounting volume name="TestVolSuspend@sharedVmfs-0" <----------

**Comment by ashahi1** [comment]
@govint Please see the following steps and their corresponding output. This still looks like a bug as we are not able to remove volume after 'Reset' of a vm is finished. User will be stuck with a left-over volume which cannot be removed.

Fresh logs have been attached at the end.

1. Created a volume.
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create -d vsphere:latest --name reproVol
reproVol
root@sc-rdops-vm02-dhcp-52-237:~#
```
2. Docker volume ls
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      SS_Volume1@sharedVmfs-0
vsphere:latest      TestVolSS@sharedVmfs-0
vsphere:latest      VolSS1@sharedVmfs-0
vsphere:latest      reproVol@sharedVmfs-0
```
3. Ran a container with volume mounted and wrote some data.
```
root@sc-rdops-vm02-dhcp-52-237:~# docker run -it --volume-driver=vsphere:latest -v reproVol@sharedVmfs-0:/vol1 --name ubReset ubuntu
root@3e45e4677fbb:/#
root@3e45e4677fbb:/# echo "Hello" > vol1/hellos2
root@3e45e4677fbb:/# cat vol1/hellos2
Hello
root@3e45e4677fbb:/#
```
4. Do 'Reset' of vm from VC.

5. After vm comes back up, list the containers. ```docker ps -a``` did list the container.
```
root@sc-rdops-vm02-dhcp-52-237:~# docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~# docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                            PORTS               NAMES
3e45e4677fbb        ubuntu              "/bin/bash"         2 minutes ago       Exited (255) About a minute ago                       ubReset
root@sc-rdops-vm02-dhcp-52-237:~#
```
6. Removed the container
```
root@sc-rdops-vm02-dhcp-52-237:~# docker rm 3e45e4677fbb
3e45e4677fbb
root@sc-rdops-vm02-dhcp-52-237:~# docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```
7. Docker volume inspect still showed volume as attached
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume inspect reproVol@sharedVmfs-0
[
    {
        "Driver": "vsphere:latest",
        "Labels": null,
        "Mountpoint": "/mnt/vmdk/reproVol@sharedVmfs-0",
        "Name": "reproVol@sharedVmfs-0",
        "Options": {},
        "Scope": "global",
        "Status": {
            "access": "read-write",
            "attach-as": "independent_persistent",
            "attached to VM": "ubuntu-VM0.3",
            "attachedVMDevice": {
                "ControllerPciSlotNumber": "160",
                "Unit": "0"
            },
            "capacity": {
                "allocated": "15MB",
                "size": "100MB"
            },
            "clone-from": "None",
            "created": "Fri Jul 28 21:41:51 2017",
            "created by VM": "ubuntu-VM0.3",
            "datastore": "sharedVmfs-0",
            "diskformat": "thin",
            "fstype": "ext4",
            "status": "attached"
        }
    }
]
root@sc-rdops-vm02-dhcp-52-237:~#
```
8. Tried removing the volume but operation failed as volume is still attached
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume rm reproVol@sharedVmfs-0
Error response from daemon: unable to remove volume: remove reproVol@sharedVmfs-0: VolumeDriver.Remove: Failed to remove volume reproVol, in use by VM = ubuntu-VM0.3.
root@sc-rdops-vm02-dhcp-52-237:~#
```

[docker.txt](https://github.com/vmware/docker-volume-vsphere/files/1184439/docker.txt)
[docker-volume-vsphere.txt](https://github.com/vmware/docker-volume-vsphere/files/1184440/docker-volume-vsphere.txt)
[vmdk_ops.txt](https://github.com/vmware/docker-volume-vsphere/files/1184438/vmdk_ops.txt)

**Comment by govint** [comment]
This time the VMListener in the service didn't get a notification at all and hence didn't remove VMDK from the VM. Was there any change in the way reset was done? The first time it happened the VM listener got the notice and was able to detach the disk.

**Comment by govint** [comment]
Effort - as noted in #1675

**Comment by pshahzeb** [comment]
https://github.com/vmware/docker-volume-vsphere/issues/1638#issuecomment-318781952

I don't see the poweroff event received by VMlistener due to VM reset. Confirmed with vmdk_ops logs

**Comment by govint** [comment]
Yes, as noted in 1649 VMListener needs to be enhanced for more events (if thats the case) or the simpler way is to deal with this in the plugin as that allows handling all scenarios with one fix.

**Comment by govint** [comment]
Fix for #1675 will address this and doesn't need additional effort.

**Comment by govint** [comment]
Fixed via #1710
